### PR TITLE
Cleanup ChannelId handling of basic methods

### LINKED ISSUE #6371: DefaultChannelId equals, hashCode, compareTo clarification
`DefaultChannelId` chooses a hash as a random value in the constructor [1] and uses the `data` array in the comparison for `equals` [2]. This means that two `DefaultChannelId` could be `equal()` but have different `hashCode()` values which violates the interface contract of `Object#hashCode()`. 

`DefaultChannelId#compareTo()` always returns `0` [3]. The violates a "strong recommendation" on the `Comparable` interface that if `(x.compareTo(y)==0) == x.equals(y)`. Why is this?

This code has been around for a while [4] so I'm not sure if it is intentional or just cruft.

[1] https://github.com/netty/netty/blame/4.1/transport/src/main/java/io/netty/channel/DefaultChannelId.java#L181
[2] https://github.com/netty/netty/blame/4.1/transport/src/main/java/io/netty/channel/DefaultChannelId.java#L264
[3] https://github.com/netty/netty/blame/4.1/transport/src/main/java/io/netty/channel/DefaultChannelId.java#L251
[4] https://github.com/netty/netty/commit/40003ed250a41286dd73683e62737e318ab003ce

### Netty version
4.1.8

### PR DESCRIPTION
**Pull request by jasontedor**
Motiviation:

Simplify implementation of compareTo/equals/hashCode for ChannelIds.

Modifications:

We simplfy the hashCode implementation for DefaultChannelId by not making it random, but making it based on the underlying data. We fix the compareTo implementation for DefaultChannelId by using lexicographic comparison of the underlying data array. We fix the compareTo implementation for CustomChannelId to avoid the possibility of overflow.

Result:

Cleaner code that is easier to maintain.

Closes #6371

### CODE REVIEW

File: transport/src/main/java/io/netty/channel/ChannelId.java [hunk 1]
```diff
@@ -43,7 +43,7 @@
  * </ul>
  * </p>
  */
-public interface ChannelId extends Serializable, Comparable<ChannelId> {
+public interface ChannelId<T extends ChannelId> extends Serializable, Comparable<T> {
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/ChannelId.java]: this is an API change. I don't think we can do this in a minor release. Can you revert this change please?

File: transport/src/main/java/io/netty/channel/DefaultChannelId.java [hunk 2]
```diff
@@ -247,25 +247,36 @@ public int hashCode() {
...
+        if (this == o) {
+            // short circuit
+            return 0;
+        }
+        // lexicographic comparison
+        int len1 = data.length;
+        int len2 = o.data.length;
+        int len = Math.min(len1, len2);
+
+        int k = 0;
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/DefaultChannelId.java]: consider just using a for loop `for (int k = 0; k < len; ++k)`

File: transport/src/main/java/io/netty/channel/DefaultChannelId.java [hunk 3]
```diff
@@ -247,25 +247,36 @@ public int hashCode() {
...
-
-        return Arrays.equals(data, ((DefaultChannelId) obj).data);
+        return this == obj || (obj instanceof DefaultChannelId && Arrays.equals(data, ((DefaultChannelId) obj).data));
     }
 
     @Override
     public String toString() {
         return asShortText();
     }
+
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/DefaultChannelId.java]: nit: kill line?

File: transport/src/main/java/io/netty/channel/embedded/EmbeddedChannelId.java [hunk 4]
```diff
@@ -62,4 +58,5 @@ public boolean equals(Object obj) {
     public String toString() {
         return "embedded";
     }
+
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/embedded/EmbeddedChannelId.java]: nit: kill line?

File: transport/src/main/java/io/netty/channel/DefaultChannelId.java [hunk 5]
```diff
@@ -247,25 +247,36 @@ public int hashCode() {
...
+        int len1 = data.length;
+        int len2 = o.data.length;
+        int len = Math.min(len1, len2);
+
+        int k = 0;
+        while (k < len) {
+            byte x = data[k];
+            byte y = o.data[k];
+            if (x != y) {
+                return x - y;
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/DefaultChannelId.java]: consider adding a comment that overflow won't occur because the arithmetic is promoted to `int`.

File: transport/src/test/java/io/netty/channel/embedded/CustomChannelId.java [hunk 6]
```diff
@@ -17,21 +17,20 @@
...
 
     @Override
-    public int compareTo(ChannelId o) {
-        if (o instanceof CustomChannelId) {
-            return id - ((CustomChannelId) o).id;
-        }
-        return -1;
+    public int compareTo(CustomChannelId o) {
+        // do not subtract for comparison, it could overflow
+        return id < o.id ? -1 : (id == o.id ? 0 : 1);
```
Comment (Scottmitch) [inline transport/src/test/java/io/netty/channel/embedded/CustomChannelId.java]: consider adding a `compare` to MathUtil for `int`. That way when we upgrade our java support it will be easier to identify and replace with the JDK methods.

**Comment by fenik17** [comment]
IMHO, for  lexicographic comparison more suitable compare bytes as unsigned.

File: transport/src/main/java/io/netty/channel/ChannelId.java (hunk 1 above)
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/ChannelId.java]: +1 .. Please revert

**Comment by jasontedor** [comment]
Thanks for the feedback @Scottmitch and @normanmaurer. I have pushed a commit addressing your comments. I also agree with the suggestion of @fenik17 to do the lexicographic comparison treating the bytes as unsigned, so I included that in that commit as well.

File: transport/src/main/java/io/netty/channel/embedded/EmbeddedChannelId.java [hunk 7]
```diff
@@ -40,12 +40,12 @@ public String asLongText() {
     }
 
     @Override
-    public int compareTo(ChannelId o) {
+    public int compareTo(final ChannelId o) {
         if (o instanceof EmbeddedChannelId) {
             return 0;
+        } else {
+            return asLongText().compareTo(o.asLongText());
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/embedded/EmbeddedChannelId.java]: please revert this as the `else`is not needed or make this:

```java
return o instanceof EmbeddedChannelId ? 0 : asLongText().compareTo(o.asLongText());
```

File: transport/src/test/java/io/netty/channel/embedded/CustomChannelId.java [hunk 8]
```diff
@@ -16,22 +16,26 @@
...
     }
 
     @Override
-    public int compareTo(ChannelId o) {
+    public int compareTo(final ChannelId o) {
         if (o instanceof CustomChannelId) {
-            return id - ((CustomChannelId) o).id;
+            final int otherId = ((CustomChannelId)o).id;
+            return MathUtil.compare(id, otherId);
+        } else {
```
Comment (normanmaurer) [inline transport/src/test/java/io/netty/channel/embedded/CustomChannelId.java]: `} else {` is not needed. Please remove

**Comment by jasontedor** [comment]
Thanks again @normanmaurer, I pushed another commit.

File: transport/src/main/java/io/netty/channel/DefaultChannelId.java [hunk 9]
```diff
@@ -247,21 +247,35 @@ public int hashCode() {
...
+            for (int k = 0; k < len; k++) {
+                byte x = data[k];
+                byte y = otherData[k];
+                if (x != y) {
+                    // treat these as unsigned bytes for comparison
+                    return (x & 0xff) - (y & 0xff);
+                }
+            }
+            return len1 - len2;
+        } else {
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/DefaultChannelId.java]: remove else block

File: transport/src/main/java/io/netty/channel/embedded/EmbeddedChannelId.java [hunk 10]
```diff
@@ -40,7 +40,7 @@ public String asLongText() {
     }
 
     @Override
-    public int compareTo(ChannelId o) {
+    public int compareTo(final ChannelId o) {
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/embedded/EmbeddedChannelId.java]: revert ?

File: transport/src/test/java/io/netty/channel/embedded/CustomChannelId.java [hunk 11]
```diff
@@ -16,22 +16,26 @@
...
         this.id = id;
     }
 
     @Override
-    public int compareTo(ChannelId o) {
+    public int compareTo(final ChannelId o) {
         if (o instanceof CustomChannelId) {
-            return id - ((CustomChannelId) o).id;
+            final int otherId = ((CustomChannelId)o).id;
+            return MathUtil.compare(id, otherId);
```
Comment (normanmaurer) [inline transport/src/test/java/io/netty/channel/embedded/CustomChannelId.java]: just merge these two lines ?

Comment (jasontedor) [inline transport/src/test/java/io/netty/channel/embedded/CustomChannelId.java]: I pushed a commit that does this.

File: transport/src/main/java/io/netty/channel/embedded/EmbeddedChannelId.java (hunk 10 above)
Comment (jasontedor) [inline transport/src/main/java/io/netty/channel/embedded/EmbeddedChannelId.java]: Why? The `final` is fine here. This is an improvement that breaks nothing.

File: transport/src/main/java/io/netty/channel/DefaultChannelId.java (hunk 9 above)
Comment (jasontedor) [inline transport/src/main/java/io/netty/channel/DefaultChannelId.java]: I pushed a commit that does this.

**Comment by jasontedor** [comment]
Thanks @normanmaurer. I've responded to your latest round of feedback.

File: transport/src/main/java/io/netty/channel/embedded/EmbeddedChannelId.java (hunk 10 above)
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/embedded/EmbeddedChannelId.java]: Just because it not really adds anything to the patch IMHO... that said I don't care too much

Comment (jasontedor) [inline transport/src/main/java/io/netty/channel/embedded/EmbeddedChannelId.java]: The goal of the patch was to clean up equals/compareTo/hashCode methods on the `ChannelId` implementations. :smile:

File: common/src/main/java/io/netty/util/internal/MathUtil.java [hunk 12]
```diff
@@ -65,6 +65,20 @@ public static boolean isOutOfBounds(int index, int length, int capacity) {
...
     /**
+     * Compares two {@code int} values.
+     *
+     * @param  x the first {@code int} to compare
+     * @param  y the second {@code int} to compare
+     * @return the value {@code 0} if {@code x == y};
+     *         {@code -1} if {@code x < y}; and
+     *         {@code 1} if {@code x > y}
+     */
+    public static int compare(int x, int y) {
```
Comment (johnou) [inline common/src/main/java/io/netty/util/internal/MathUtil.java]: do we need to introduce this? can't we use java.lang.Integer#compare?

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/MathUtil.java]: We can't as we need to still support java6

Comment (johnou) [inline common/src/main/java/io/netty/util/internal/MathUtil.java]: ow, for some reason I thought Netty 4.0.x was java6 and Netty 4.1.x was java7.

Comment (jasontedor) [inline common/src/main/java/io/netty/util/internal/MathUtil.java]: No, it's 1.6 in the [pom.xml](https://github.com/netty/netty/blob/c92565d5c7000d2bd74355fe18cbdf98bfc5ba0d/pom.xml#L625).

**Comment by normanmaurer** [comment]
@Scottmitch PTAL and pull in when satisfied. 

@jasontedor thanks again. Great work!

**Comment by jasontedor** [comment]
Thanks @normanmaurer. I squashed.

**Comment by Scottmitch** [comment]
4.1 (d59b484)

Thanks @jasontedor !

**Comment by Scottmitch** [comment]
Also 4.1 (544d771)

File: transport/src/main/java/io/netty/channel/embedded/EmbeddedChannelId.java (hunk 10 above)
Comment (trustin) [inline transport/src/main/java/io/netty/channel/embedded/EmbeddedChannelId.java]: I think it's always good to clean the code up while making modifications as long as it doesn't drain too much cognitive load. :-)
